web_syncer/models/syncer.py

```diff
@@ -0,0 +1,353 @@
+            return self._fields[fn].type == 'integer' and fn == 'res_id'
+
+        def is_fake_one2many_for_model(model, rec):
+            for field_name in ['res_model', 'model']:
+                if field_name in rec and rec[field_name] == model:
+                    return True
+            return False
+
+        for x2m_field in self._syncer_satellites:
+            model = x2m_field['model']
+            field_name = x2m_field['field_name']
+
+            if x2m_field['type'] == 'one2many':
+                for rec in self:
+                    for inverse_name in x2m_field['inverse_name']:
+
+                        # Special case for fake one2many field
+                        #  (i.e. ir.attachments)
+                        if is_fake_one2many(inverse_name):
+                            if not is_fake_one2many_for_model(model, rec):
+                                continue
+                            record = self.env[model].browse(rec[inverse_name])
+                        else:
+                            record = rec[inverse_name]
+                        if not record:
+                            continue
+                        self.push_indirect_notification(model, record.id, {
+                            'method': 'write',
+                            'record_name': record.name_get()[0][1],
+                            'data': {field_name: record[field_name].ids},
+                        })
+
+            elif x2m_field['type'] == 'many2many':
+                for rec in self:
+                    for record in rec[x2m_field['inverse_field']]:
+                        self.push_indirect_notification(model, record.id, {
+                            'method': 'write',
+                            'record_name': record.name_get()[0][1],
+                            'data': {field_name: record[field_name].ids},
+                        })
+
+    def process_x2m_notifications_before_write(self, vals):
+        for x2m_field in self._syncer_satellites:
+            if x2m_field['type'] != 'one2many':
+                continue
+
+            processed = {}
+            for inverse_name in x2m_field['inverse_name']:
+                if inverse_name not in vals:
+                    continue
+
+                record_id = vals[inverse_name]
+                if not record_id:
+                    continue
+
+                model = x2m_field['model']
+                field_name = x2m_field['field_name']
+
+                if not isinstance(record_id, int):
+                    record = record_id
+                else:
+                    record = self.env[model].with_context(
+                        prefetch_fields=[field_name]
+                    ).browse(record_id)
+
+                for rec in self:
+                    original = rec[inverse_name]
+                    if not original:
+                        continue
+
+                    # Parent has changed and we need to notify previous parent
```

Comment:
Got some issues here in cases where trying to convert a lead to a task for example or merging contacts...
Tested in v10, but same code applies:

```
Traceback (most recent call last):
  File "/home/odoo/FULLOCA/__src__/OCB/odoo/http.py", line 641, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/odoo/FULLOCA/__src__/OCB/odoo/http.py", line 683, in dispatch
    result = self._call_function(**self.params)
  File "/home/odoo/FULLOCA/__src__/OCB/odoo/http.py", line 333, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/odoo/FULLOCA/__src__/OCB/odoo/service/model.py", line 101, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/odoo/FULLOCA/__src__/OCB/odoo/http.py", line 326, in checked_call
    result = self.endpoint(*a, **kw)
  File "/home/odoo/xrepo/moodular/web_syncer/http.py", line 9, in __call__
    result = self.method(*args, **kw)
  File "/home/odoo/FULLOCA/__src__/OCB/odoo/http.py", line 506, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/FULLOCA/__src__/OCB/addons/web/controllers/main.py", line 894, in call_button
    action = self._call_kw(model, method, args, {})
  File "/home/odoo/FULLOCA/__src__/OCB/addons/web/controllers/main.py", line 882, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/odoo/FULLOCA/__src__/OCB/odoo/api.py", line 689, in call_kw
    return call_kw_multi(method, model, args, kwargs)
  File "/home/odoo/FULLOCA/__src__/OCB/odoo/api.py", line 680, in call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/odoo/FULLOCA/__src__/OCB/addons/crm/wizard/base_partner_merge.py", line 649, in action_merge
    self._merge(self.partner_ids.ids, self.dst_partner_id)
  File "/home/odoo/FULLOCA/__src__/OCB/addons/crm/wizard/base_partner_merge.py", line 327, in _merge
    self._update_reference_fields(src_partners, dst_partner)
  File "/home/odoo/FULLOCA/__src__/OCB/addons/crm/wizard/base_partner_merge.py", line 231, in _update_reference_fields
    update_records('mail.followers', src=partner, field_model='res_model')
  File "/home/odoo/FULLOCA/__src__/OCB/addons/crm/wizard/base_partner_merge.py", line 220, in update_records
    return records.sudo().write({field_id: dst_partner.id})
  File "/home/odoo/FULLOCA/__src__/OCB/addons/mail/models/mail_followers.py", line 133, in write
    res = super(Followers, self).write(vals)
  File "/home/odoo/xrepo/moodular/web_syncer/models/syncer.py", line 95, in write
    self.process_x2m_notifications_before_write(vals)
  File "/home/odoo/xrepo/moodular/web_syncer/models/syncer.py", line 201, in process_x2m_notifications_before_write
    if original.id != record.id:
AttributeError: 'int' object has no attribute 'id'
```

Reply:
This is fixed now.